File: kubefn-runtime-python/kubefn/introspection.py
"""
Causal State Introspection — request tracing and event capture.

Simplified Python equivalent of the JVM CausalCaptureEngine.
Records structured events (request lifecycle, function invocations,
heap mutations) into a bounded ring buffer for diagnostics.
"""

import enum
import threading
import time
import uuid
from dataclasses import dataclass, field


class EventType(enum.Enum):
    REQUEST_START = "REQUEST_START"
    REQUEST_END = "REQUEST_END"
    FUNCTION_START = "FUNCTION_START"
    FUNCTION_END = "FUNCTION_END"
    HEAP_PUBLISH = "HEAP_PUBLISH"
    HEAP_GET_HIT = "HEAP_GET_HIT"
    HEAP_GET_MISS = "HEAP_GET_MISS"
    ERROR = "ERROR"


@dataclass(frozen=True, slots=True)
class CausalEvent:
    """An immutable event in the causal trace."""
    event_id: str
    timestamp: float
    request_id: str
    event_type: EventType
    function_name: str
    group_name: str
    heap_key: str | None = None
    duration_ns: int | None = None
    detail: str | None = None


class EventRing:
    """
    Bounded ring buffer for causal events.
    Thread-safe. Oldest events are evicted when capacity is reached.
    """

    def __init__(self, capacity: int = 50_000):
        self._capacity = capacity
        self._events: list[CausalEvent] = []
        self._lock = threading.Lock()

    def append(self, event: CausalEvent) -> None:
        with self._lock:
            self._events.append(event)
            if len(self._events) > self._capacity:
                # Drop oldest 10% to avoid frequent trimming
                drop_count = self._capacity // 10
                self._events = self._events[drop_count:]

    def get_by_request_id(self, request_id: str) -> list[CausalEvent]:
        with self._lock:
            return [e for e in self._events if e.request_id == request_id]

    def recent(self, limit: int = 100) -> list[CausalEvent]:
        with self._lock:
            return list(self._events[-limit:])

    def size(self) -> int:
        return len(self._events)

    def clear(self) -> None:
        with self._lock:
            self._events.clear()


class TraceAssembler:
    """Builds a structured trace dict from a list of causal events."""

    @staticmethod
    def assemble(events: list[CausalEvent]) -> dict:
        if not events:
            return {"error": "no events found"}

        request_id = events[0].request_id
        steps: list[dict] = []
        heap_mutations: list[dict] = []
        errors: list[dict] = []
        total_duration_ns: int | None = None

        for event in events:
            match event.event_type:
                case EventType.REQUEST_START:
                    pass  # captured in top-level metadata
                case EventType.REQUEST_END:
                    total_duration_ns = event.duration_ns
                case EventType.FUNCTION_START:
                    steps.append({
                        "type": "function_start",
                        "function": event.function_name,
                        "group": event.group_name,
                        "timestamp": event.timestamp,
                    })
                case EventType.FUNCTION_END:
                    steps.append({
                        "type": "function_end",
                        "function": event.function_name,
                        "group": event.group_name,
                        "durationNs": event.duration_ns,
                        "timestamp": event.timestamp,
                    })
                case EventType.HEAP_PUBLISH:
                    heap_mutations.append({
                        "action": "publish",
                        "key": event.heap_key,
                        "function": event.function_name,
                        "group": event.group_name,
                        "timestamp": event.timestamp,
                    })
                case EventType.HEAP_GET_HIT:
                    heap_mutations.append({
                        "action": "get_hit",
                        "key": event.heap_key,
                        "function": event.function_name,
                        "timestamp": event.timestamp,
                    })
                case EventType.HEAP_GET_MISS:
                    heap_mutations.append({
                        "action": "get_miss",
                        "key": event.heap_key,
                        "function": event.function_name,
                        "timestamp": event.timestamp,
                    })
                case EventType.ERROR:
                    errors.append({
                        "function": event.function_name,
                        "detail": event.detail,
                        "timestamp": event.timestamp,
                    })

        trace: dict = {
            "requestId": request_id,
            "steps": steps,
            "heapMutations": heap_mutations,
            "stepCount": len(steps),
        }
        if total_duration_ns is not None:
            trace["totalDurationMs"] = round(total_duration_ns / 1_000_000, 3)
        if errors:
            trace["errors"] = errors

        return trace


class CaptureEngine:
    """
    Singleton-style capture engine for causal introspection.
    Records events and provides query methods for the admin API.
    """

    def __init__(self, capacity: int = 50_000):
        self._ring = EventRing(capacity=capacity)
        # Track recent request IDs for quick lookup
        self._recent_request_ids: list[str] = []
        self._recent_lock = threading.Lock()
        self._max_recent = 500

    def _emit(self, event: CausalEvent) -> None:
        self._ring.append(event)

    def _track_request_id(self, request_id: str) -> None:
        with self._recent_lock:
            self._recent_request_ids.append(request_id)
            if len(self._recent_request_ids) > self._max_recent:
                self._recent_request_ids = self._recent_request_ids[
                    -self._max_recent :
                ]

    @staticmethod
    def _new_event_id() -> str:
        return uuid.uuid4().hex[:16]

    # ── Capture methods ───────────────────────────────────────────────

    def capture_request_start(
        self, request_id: str, function_name: str, group_name: str
    ) -> None:
        self._track_request_id(request_id)
        self._emit(CausalEvent(
            event_id=self._new_event_id(),
            timestamp=time.time(),
            request_id=request_id,
            event_type=EventType.REQUEST_START,
            function_name=function_name,
            group_name=group_name,
        ))

    def capture_request_end(
        self,
        request_id: str,
        function_name: str,
        group_name: str,
        duration_ns: int,
    ) -> None:
        self._emit(CausalEvent(
            event_id=self._new_event_id(),
            timestamp=time.time(),
            request_id=request_id,
            event_type=EventType.REQUEST_END,
            function_name=function_name,
            group_name=group_name,
            duration_ns=duration_ns,
        ))

    def capture_function_start(
        self, request_id: str, function_name: str, group_name: str
    ) -> None:
        self._emit(CausalEvent(
            event_id=self._new_event_id(),
            timestamp=time.time(),
            request_id=request_id,
            event_type=EventType.FUNCTION_START,
            function_name=function_name,
            group_name=group_name,
        ))

    def capture_function_end(
        self,
        request_id: str,
        function_name: str,
        group_name: str,
        duration_ns: int,
    ) -> None:
        self._emit(CausalEvent(
            event_id=self._new_event_id(),
            timestamp=time.time(),
            request_id=request_id,
            event_type=EventType.FUNCTION_END,
            function_name=function_name,
            group_name=group_name,
            duration_ns=duration_ns,
        ))

    def capture_heap_publish(
        self,
        request_id: str,
        function_name: str,
        group_name: str,
        heap_key: str,
    ) -> None:
        self._emit(CausalEvent(
            event_id=self._new_event_id(),
            timestamp=time.time(),
            request_id=request_id,
            event_type=EventType.HEAP_PUBLISH,
            function_name=function_name,
            group_name=group_name,
            heap_key=heap_key,
        ))

    def capture_heap_get(
        self,
        request_id: str,
        function_name: str,
        group_name: str,
        heap_key: str,
        hit: bool,
    ) -> None:
        self._emit(CausalEvent(
            event_id=self._new_event_id(),
            timestamp=time.time(),
            request_id=request_id,
            event_type=EventType.HEAP_GET_HIT if hit else EventType.HEAP_GET_MISS,
            function_name=function_name,
            group_name=group_name,
            heap_key=heap_key,
        ))

    def capture_error(
        self,
        request_id: str,
        function_name: str,
        group_name: str,
        detail: str,
    ) -> None:
        self._emit(CausalEvent(
            event_id=self._new_event_id(),
            timestamp=time.time(),
            request_id=request_id,
            event_type=EventType.ERROR,
            function_name=function_name,
            group_name=group_name,
            detail=detail,
        ))

    # ── Query methods ─────────────────────────────────────────────────

    def get_trace(self, request_id: str) -> dict:
        events = self._ring.get_by_request_id(request_id)
        return TraceAssembler.assemble(events)

    def recent_traces(self, limit: int = 20) -> list[dict]:
        with self._recent_lock:
            request_ids = list(reversed(self._recent_request_ids[-limit:]))

        traces = []
        for rid in request_ids:
            events = self._ring.get_by_request_id(rid)
            if events:
                traces.append(TraceAssembler.assemble(events))
        return traces

    def event_count(self) -> int:
        return self._ring.size()

File: kubefn-runtime-node/src/introspection.js
/**
 * Introspection — causal event capture, ring buffer storage, trace assembly.
 *
 * Every function invocation, heap operation, and request lifecycle event is
 * captured as a CausalEvent and stored in a fixed-capacity ring buffer.
 * Traces are assembled on demand from events sharing a requestId.
 */

import { randomUUID } from 'node:crypto';

/** Canonical event types. */
export const EventType = Object.freeze({
  REQUEST_START: 'REQUEST_START',
  REQUEST_END: 'REQUEST_END',
  FUNCTION_START: 'FUNCTION_START',
  FUNCTION_END: 'FUNCTION_END',
  HEAP_PUBLISH: 'HEAP_PUBLISH',
  HEAP_GET: 'HEAP_GET',
  HEAP_MISS: 'HEAP_MISS',
  HEAP_REMOVE: 'HEAP_REMOVE',
  CIRCUIT_OPEN: 'CIRCUIT_OPEN',
  TIMEOUT: 'TIMEOUT',
  ERROR: 'ERROR',
});

export class CausalEvent {
  /**
   * @param {object} fields
   * @param {string} fields.requestId
   * @param {string} fields.eventType    - one of EventType values
   * @param {string} [fields.functionName]
   * @param {string} [fields.heapKey]
   * @param {number} [fields.durationNs]
   * @param {string} [fields.detail]
   */
  constructor({ requestId, eventType, functionName, heapKey, durationNs, detail }) {
    this.eventId = randomUUID();
    this.timestamp = Date.now();
    this.requestId = requestId;
    this.eventType = eventType;
    this.functionName = functionName || null;
    this.heapKey = heapKey || null;
    this.durationNs = durationNs ?? null;
    this.detail = detail || null;
  }
}

/**
 * Fixed-capacity ring buffer for CausalEvent instances.
 * Overwrites oldest entries when full — O(1) append.
 */
export class EventRing {
  /**
   * @param {number} capacity - max events stored (default 50 000)
   */
  constructor(capacity = 50_000) {
    this._capacity = capacity;
    this._buffer = new Array(capacity);
    this._head = 0;   // next write position
    this._size = 0;
  }

  /** Append an event to the ring. */
  append(event) {
    this._buffer[this._head] = event;
    this._head = (this._head + 1) % this._capacity;
    if (this._size < this._capacity) this._size++;
  }

  /**
   * Return all events for a given requestId (scan — tolerable for traces).
   * @param {string} requestId
   * @returns {CausalEvent[]}
   */
  getByRequestId(requestId) {
    const result = [];
    const start = this._size < this._capacity ? 0 : this._head;
    for (let i = 0; i < this._size; i++) {
      const idx = (start + i) % this._capacity;
      const evt = this._buffer[idx];
      if (evt && evt.requestId === requestId) {
        result.push(evt);
      }
    }
    return result;
  }

  /**
   * Return the N most recent events.
   * @param {number} limit
   * @returns {CausalEvent[]}
   */
  recent(limit = 100) {
    const n = Math.min(limit, this._size);
    const result = new Array(n);
    for (let i = 0; i < n; i++) {
      const idx = (this._head - 1 - i + this._capacity) % this._capacity;
      result[i] = this._buffer[idx];
    }
    return result;
  }

  get size() {
    return this._size;
  }
}

/** Assembles a human-readable trace from raw events. */
export class TraceAssembler {
  /**
   * @param {string} requestId
   * @param {CausalEvent[]} events - events sorted by timestamp
   * @returns {object}
   */
  assemble(requestId, events) {
    if (!events || events.length === 0) {
      return { requestId, found: false, events: [] };
    }

    const sorted = [...events].sort((a, b) => a.timestamp - b.timestamp);
    const first = sorted[0];
    const last = sorted[sorted.length - 1];

    const heapOps = sorted.filter(
      (e) => e.eventType === EventType.HEAP_PUBLISH ||
             e.eventType === EventType.HEAP_GET ||
             e.eventType === EventType.HEAP_MISS ||
             e.eventType === EventType.HEAP_REMOVE
    );

    const errors = sorted.filter(
      (e) => e.eventType === EventType.ERROR || e.eventType === EventType.TIMEOUT
    );

    const reqEnd = sorted.find((e) => e.eventType === EventType.REQUEST_END);

    return {
      requestId,
      found: true,
      startTime: first.timestamp,
      endTime: last.timestamp,
      durationMs: last.timestamp - first.timestamp,
      functionName: first.functionName,
      totalDurationNs: reqEnd ? reqEnd.durationNs : null,
      eventCount: sorted.length,
      heapOps: heapOps.length,
      errors: errors.length,
      events: sorted,
    };
  }
}

/**
 * High-level capture API used by the server and heap-exchange.
 * Delegates to EventRing for storage and TraceAssembler for queries.
 */
export class CaptureEngine {
  /**
   * @param {number} [ringCapacity=50000]
   */
  constructor(ringCapacity = 50_000) {
    this._ring = new EventRing(ringCapacity);
    this._assembler = new TraceAssembler();
  }

  // ── lifecycle events ──────────────────────────────────────────

  captureRequestStart(requestId, funcName) {
    this._ring.append(new CausalEvent({
      requestId,
      eventType: EventType.REQUEST_START,
      functionName: funcName,
    }));
  }

  captureRequestEnd(requestId, durationNs, error) {
    this._ring.append(new CausalEvent({
      requestId,
      eventType: EventType.REQUEST_END,
      durationNs,
      detail: error ? String(error) : null,
    }));
  }

  captureFunctionStart(requestId, funcName) {
    this._ring.append(new CausalEvent({
      requestId,
      eventType: EventType.FUNCTION_START,
      functionName: funcName,
    }));
  }

  captureFunctionEnd(requestId, funcName, durationNs, error) {
    this._ring.append(new CausalEvent({
      requestId,
      eventType: EventType.FUNCTION_END,
      functionName: funcName,
      durationNs,
      detail: error ? String(error) : null,
    }));
  }

  // ── heap events ───────────────────────────────────────────────

  captureHeapPublish(requestId, key) {
    this._ring.append(new CausalEvent({
      requestId,
      eventType: EventType.HEAP_PUBLISH,
      heapKey: key,
    }));
  }

  captureHeapGet(requestId, key, hit) {
    this._ring.append(new CausalEvent({
      requestId,
      eventType: hit ? EventType.HEAP_GET : EventType.HEAP_MISS,
      heapKey: key,
    }));
  }

  // ── query API ─────────────────────────────────────────────────

  /**
   * Assemble a full trace for a request.
   * @param {string} requestId
   * @returns {object}
   */
  getTrace(requestId) {
    const events = this._ring.getByRequestId(requestId);
    return this._assembler.assemble(requestId, events);
  }

  /**
   * Return the N most recent traces (grouped by requestId).
   * @param {number} limit
   * @returns {object[]}
   */
  recentTraces(limit = 20) {
    const recent = this._ring.recent(limit * 10); // over-sample to find enough unique requests
    const seen = new Map();

    for (const evt of recent) {
      if (evt && evt.requestId && !seen.has(evt.requestId)) {
        seen.set(evt.requestId, true);
        if (seen.size >= limit) break;
      }
    }

    const traces = [];
    for (const reqId of seen.keys()) {
      traces.push(this.getTrace(reqId));
    }
    return traces;
  }

  get ringSize() {
    return this._ring.size;
  }
}
